components/core/src/clp/streaming_archive/single_file_archive/writer.cpp

```diff
@@ -0,0 +1,251 @@
+        -> std::vector<FileInfo> {
+    std::vector<FileInfo> files;
+    uint64_t offset = 0;
+
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        files.emplace_back(FileInfo{std::string(static_archive_file_name), offset});
+        get_file_size_and_update_offset(multi_file_archive_path / static_archive_file_name, offset);
+    }
+
+    std::filesystem::path segment_dir_path = multi_file_archive_path / cSegmentsDirname;
+
+    for (size_t i = 0; i < next_segment_id; ++i) {
+        auto const segment_id = std::to_string(i);
+        files.emplace_back(FileInfo{segment_id, offset});
+        get_file_size_and_update_offset(segment_dir_path / segment_id, offset);
+    }
+
+    // Add sentinel indicating total size of all files.
+    files.emplace_back(FileInfo{"", offset});
+
+    return files;
+}
+
+auto pack_single_file_archive_metadata(
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> std::stringstream {
+    SingleFileArchiveMetadata single_file_archive{
+            .archive_files = get_file_infos(multi_file_archive_path, next_segment_id),
+            .num_segments = next_segment_id,
+    };
+
+    std::stringstream buf;
+    msgpack::pack(buf, single_file_archive);
+
+    return buf;
+}
+
+auto write_archive_header(FileWriter& single_file_archive_writer, size_t packed_metadata_size)
+        -> void {
+    SingleFileArchiveHeader header{
+            .magic = cUnstructuredSfaMagicNumber,
+            .version = cVersion,
+            .metadata_size = packed_metadata_size,
+            .unused{}
+    };
+
+    single_file_archive_writer.write(reinterpret_cast<char const*>(&header), sizeof(header));
+}
+
+auto write_archive_metadata(
+        FileWriter& single_file_archive_writer,
+        std::stringstream const& packed_metadata
+) -> void {
+    single_file_archive_writer.write(packed_metadata.str().data(), packed_metadata.str().size());
+}
+
+auto
+write_archive_file(std::filesystem::path const& file_path, FileWriter& single_file_archive_writer)
+        -> void {
+    FileReader reader(file_path.string());
+    std::array<char, cReadBlockSize> read_buffer{};
+    while (true) {
+        size_t num_bytes_read{};
+        ErrorCode const error_code
+                = reader.try_read(read_buffer.data(), cReadBlockSize, num_bytes_read);
+        if (ErrorCode_EndOfFile == error_code) {
+            break;
+        }
+        if (ErrorCode_Success != error_code) {
+            throw OperationFailed(error_code, __FILENAME__, __LINE__);
+        }
+        single_file_archive_writer.write(read_buffer.data(), num_bytes_read);
+    }
+}
+
+auto write_archive_files(
+        FileWriter& single_file_archive_writer,
+        std::filesystem::path const& multi_file_archive_path,
+        segment_id_t next_segment_id
+) -> void {
+    for (auto const& static_archive_file_name : cStaticArchiveFileNames) {
+        std::filesystem::path static_archive_file_path
```

Comment:
auto const?
```suggestion
        auto const static_archive_file_path
```
also can we use {} instead of = for initialization? not sure if auto will work properly here